### PR TITLE
[RW Separation] Introduce allocation filter to control placement of search only replicas

### PR DESCRIPTION
**Pull request by mch2**
### Description
Introduce allocation filter to control placement of search only replicas.

The new filter `cluster.routing.allocation.search.replica.dedicated.include` acts as a normal include filter in that you can specify multiple values and search shards will allocate there if one of the attributes is present on a particular node.  However, it is different from existing include filters in that it is exclusive to search replicas and will exclude any non search shard for all of the attributes set.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/15445

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/SearchReplicaAllocationDecider.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/decider/SearchReplicaAllocationDecider.java:31]: You don't need these headers

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/SearchReplicaAllocationDecider.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                    Decision.NO,
+                    NAME,
+                    "only search replicas can be allocated to node with setting [%s] filters [%s]",
+                    SEARCH_REPLICA_ROUTING_INCLUDE_GROUP_PREFIX,
+                    searchReplicaIncludeFilters
+                );
+            }
+        }
+        return null;
+    }
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/decider/SearchReplicaAllocationDecider.java:144]: Style nitpick, but I think you can avoid the null value by merging these two methods:

```suggestion
    private Decision shouldFilter(ShardRouting shardRouting, DiscoveryNode node, RoutingAllocation allocation) {
        if (searchReplicaIncludeFilters != null) {
            final boolean match = searchReplicaIncludeFilters.match(node);
            if (match == false && routing.isSearchOnly()) {
                return allocation.decision(
                    Decision.NO,
                    NAME,
                    "node does not match shard setting [%s] filters [%s]",
                    SEARCH_REPLICA_ROUTING_INCLUDE_GROUP_PREFIX,
                    searchReplicaIncludeFilters
                );
            }
            // filter will only apply to search replicas
            if (routing.isSearchOnly() == false && match) {
                return allocation.decision(
                    Decision.NO,
                    NAME,
                    "only search replicas can be allocated to node with setting [%s] filters [%s]",
                    SEARCH_REPLICA_ROUTING_INCLUDE_GROUP_PREFIX,
                    searchReplicaIncludeFilters
                );
            }
        }
        return allocation.decision(Decision.YES, NAME, "node passes include/exclude/require filters");
    }
```

**Comment by mch2** [comment]
> ❌ Gradle check result for [02977be](https://github.com/opensearch-project/OpenSearch/commit/02977be573d00446fbc4801dd69c3480518600fd): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/46112/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/15584

**Review by andrross — APPROVED** [review]

**Comment by mch2** [comment]
removed changelog entry here to reduce all the backport / gc pain, will include 1 entry for all the separation prs and link to the design issue.

**Comment by mch2** [comment]
> FAILURE

https://github.com/opensearch-project/OpenSearch/issues/15692

**Comment by mch2** [comment]
> ❌ Gradle check result for [05d8e19](https://github.com/opensearch-project/OpenSearch/commit/05d8e19fe69f63d33375620e1d4fe719f15c0bb9):
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/14328

**Pull request merged**